<a href="https://colab.research.google.com/github/dennismathewjose/Financial-ChatBot--BCG-Gen-AI-Forage-/blob/main/BCG_GenAI_Financial_Analysis_10_K_reports.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Your task is to manually extract key financial data for the last three fiscal years from the 10-K filings of Microsoft, Tesla, and Apple. Following the data collection, you will use Python to analyze this data, focusing on trends and insights that could inform the development of an AI-powered financial chatbot.



# Loading and Pre Processing

In [17]:
import pandas as pd
import numpy as np
from plotly.subplots import make_subplots
import plotly.io as pio
import plotly.graph_objects as go
import plotly.express as px

custom_palette = ["#636EFA", "#3366cc","#1F77B4"]
pio.templates["custom_template"] = go.layout.Template(
    layout_colorway=custom_palette
)

pio.templates.default = "custom_template"

In [2]:
df = pd.read_csv('10-K_financial_figures.csv')
df.head()

Company  Year Total Revenue Net Income Total Assets Total Liabilities  \
0   Apple  2022       394,328     99,803      352,755           302,083   
1   Apple  2023       383,285     96,995     352,583            290,437   
2   Apple  2024       391,035     93,736      364,980          308,030    
3   Tesla  2022        81,462     12,587       82,338            36,440   
4   Tesla  2023        96,773     14,974      106,618            43,009   

  Cash Flow  
0   122,151  
1   110,543  
2   118,254  
3    14,724  
4    13,256

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Company            9 non-null      object
 1   Year               9 non-null      int64 
 2   Total Revenue      9 non-null      object
 3   Net Income         9 non-null      object
 4   Total Assets       9 non-null      object
 5   Total Liabilities  9 non-null      object
 6   Cash Flow          9 non-null      object
dtypes: int64(1), object(6)
memory usage: 636.0+ bytes


In [4]:
columns = ['Total Revenue','Net Income','Total Assets','Total Liabilities','Cash Flow']
#converting these columns to integer type
for col in columns:
    df[col] = df[col].str.replace(',','').astype(int)

# Trend and Growth Analysis

In [5]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100

In [6]:
df['Net Grwoth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [7]:
df

Company  Year  Total Revenue  Net Income  Total Assets  \
0      Apple  2022         394328       99803        352755   
1      Apple  2023         383285       96995        352583   
2      Apple  2024         391035       93736        364980   
3      Tesla  2022          81462       12587         82338   
4      Tesla  2023          96773       14974        106618   
5      Tesla  2024          97690        7153        122070   
6  Microsoft  2022         198270       72738        364840   
7  Microsoft  2023         211915       72361        411976   
8  Microsoft  2024         245122       88136        512163   

   Total Liabilities  Cash Flow  Revenue Growth (%)  Net Grwoth (%)  
0             302083     122151                 NaN             NaN  
1             290437     110543           -2.800461       -2.813543  
2             308030     118254            2.021994       -3.359967  
3              36440      14724                 NaN             NaN  
4              43009      13256           18.795267       18.964010  
5              48390      14923            0.947578      -52.230533  
6             198298      89035                 NaN             NaN  
7             205753      87582            6.882030       -0.518299  
8             243686     118548           15.669962       21.800417

In [8]:
df['Average Revenvue'] = df.groupby(['Company'])['Total Revenue'].transform('mean')

In [9]:
df

Company  Year  Total Revenue  Net Income  Total Assets  \
0      Apple  2022         394328       99803        352755   
1      Apple  2023         383285       96995        352583   
2      Apple  2024         391035       93736        364980   
3      Tesla  2022          81462       12587         82338   
4      Tesla  2023          96773       14974        106618   
5      Tesla  2024          97690        7153        122070   
6  Microsoft  2022         198270       72738        364840   
7  Microsoft  2023         211915       72361        411976   
8  Microsoft  2024         245122       88136        512163   

   Total Liabilities  Cash Flow  Revenue Growth (%)  Net Grwoth (%)  \
0             302083     122151                 NaN             NaN   
1             290437     110543           -2.800461       -2.813543   
2             308030     118254            2.021994       -3.359967   
3              36440      14724                 NaN             NaN   
4              43009      13256           18.795267       18.964010   
5              48390      14923            0.947578      -52.230533   
6             198298      89035                 NaN             NaN   
7             205753      87582            6.882030       -0.518299   
8             243686     118548           15.669962       21.800417   

   Average Revenvue  
0     389549.333333  
1     389549.333333  
2     389549.333333  
3      91975.000000  
4      91975.000000  
5      91975.000000  
6     218435.666667  
7     218435.666667  
8     218435.666667

In [10]:
df.fillna(0,inplace=True)

In [20]:
summary = df.groupby('Company').agg({
    'Revenue Growth (%)': 'mean',
    'Net Grwoth (%)': 'mean'
}).reset_index()

print("\nYear-over-Year Average Growth Rates (%):")
summary


Year-over-Year Average Growth Rates (%):


Company  Revenue Growth (%)  Net Grwoth (%)
0      Apple           -0.259489       -2.057837
1  Microsoft            7.517331        7.094040
2      Tesla            6.580948      -11.088841

## Financial Analysis Summary (2022–2024)

---

### Apple

**Revenue Growth**
- 2023: -2.80%  
- 2024: +2.02%  

**Net Income Growth**
- 2023: -2.81%  
- 2024: -3.36%  

**Observations**
- Apple shows revenue stagnation and declining profitability over the last two years.
- Total Assets rose from 352.8 billion to 365.0 billion, showing moderate growth.
- Operating cash flow remains strong, ending at 118.3 billion in 2024.
- Average Revenue over 3 years: 389.5 billion.

---

### Tesla

**Revenue Growth**
- 2023: +18.80%  
- 2024: +0.95%  

**Net Income Growth**
- 2023: +18.96%  
- 2024: -52.23%  

**Observations**
- Tesla had a sharp revenue surge in 2023, followed by stagnation in 2024.
- The massive drop in net income in 2024 is a red flag.
- Operating cash flow remained flat (approximately 14.9 billion).
- Total Assets grew significantly from 82.3 billion to 122.1 billion.
- Average Revenue over 3 years: 91.9 billion.

---

### Microsoft

**Revenue Growth**
- 2023: +6.88%  
- 2024: +15.67%  

**Net Income Growth**
- 2023: -0.52%  
- 2024: +21.80%  

**Observations**
- Microsoft exhibits steady and accelerating growth in both revenue and net income.
- Total Assets expanded from 364.8 billion to 512.2 billion.
- Operating cash flow improved from 89.0 billion to 118.5 billion, reflecting strong financial health.
- Average Revenue over 3 years: 218.4 billion.

---



# Profitability Analysis

In [16]:
df['Net Profit Margin'] = df['Net Income'] / df['Total Revenue'] * 100
#creating a bar chart to see the net profit margin of each company during each year using a bar chart using plotly.express with colors of the bar as blue gradients
fig = px.bar(df, x='Year', y='Net Profit Margin', color='Company', barmode='group',title='Net Profit Margin of each company over the year 2022-2024')
fig.show()

## Net Profit Margin Analysis (2022–2024)

**Apple**  
Apple's net profit margin remained relatively stable, around 25.4% in both 2022 and 2023, with a slight decline to approximately 24% in 2024. This reflects consistent profitability, though not the highest among the three companies.

**Tesla**  
Tesla maintained a steady margin of approximately 15.5% across 2022 and 2023, but experienced a sharp decline to about 7.3% in 2024. This significant drop suggests rising costs or declining operational efficiency.

**Microsoft**  
Microsoft consistently delivered the highest net profit margins, peaking at around 36.9% in 2022 and staying strong through 2023 (approximately 34.3%) and 2024 (approximately 35.8%). This trend underscores Microsoft's strong earnings performance and superior cost management.

**Insight**  
Among the three, Microsoft demonstrates the most robust and consistent profitability. Apple remains stable but does not lead in margin strength. Tesla's 2024 decline may require further investigation into cost structure or one-time impacts.


In [18]:
#operating cash flow margins
df['Cash Flow Margin'] = df['Cash Flow'] / df['Total Revenue'] * 100
fig = px.bar(df, x = 'Year', y = 'Cash Flow Margin', color = 'Company', barmode = 'group',title = 'Cash flow margin of each company over the year 2022-2024')
fig.show()

## Cash Flow Margin Analysis (2022–2024)

**Apple**  
Apple maintains a strong and consistent cash flow margin, ranging from approximately 29% to 31% across the three years. This indicates efficient cash generation relative to revenue.

**Tesla**  
Tesla consistently posts the lowest cash flow margins among the three, declining from around 18% in 2022 to approximately 15% in 2024. This suggests potential constraints in converting revenue into operating cash flow.

**Microsoft**  
Microsoft leads in cash flow efficiency, with margins rising from approximately 45% in 2022 to nearly 48% in 2024. This demonstrates superior operational efficiency and strong cash-generating capabilities.

**Insight**  
Microsoft consistently outperforms in cash flow margin, reflecting efficient operations and financial stability. Apple performs well with steady margins, while Tesla shows relatively weak performance in cash conversion, which may warrant further attention.


# Balance Sheet Ratios

In [19]:


df['Debt to asset ratio'] = df['Total Liabilities'] / df['Total Assets']
fig = make_subplots(rows = 1, cols = 3, specs=[[{'type':'domain'},{'type':'domain'},{'type':'domain'}]])

fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Apple']['Year'], values = df[df['Company'] == 'Apple']['Debt to asset ratio'],name='Apple', title='Apple'),
    row = 1, col = 1
)
fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Tesla']['Year'], values = df[df['Company'] == 'Tesla']['Debt to asset ratio'],name = 'Tesla',title='Tesla'),
    row = 1, col = 2
)
fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Microsoft']['Year'], values = df[df['Company'] == 'Microsoft']['Debt to asset ratio'],name = 'Microsoft',title='Mircrosoft'),
    row = 1, col = 3
)
fig.update_layout(title_text="Debt to Asset Ratio of each Company from 2022 - 2024")
fig.show()

## Debt-to-Asset Ratio Analysis (2022–2024)

**Apple**  
Apple's debt-to-asset ratio has remained relatively stable, with a slight increase from 32.6% in 2023 to 33.4% in 2024. This suggests a modest rise in liabilities relative to total assets.

**Tesla**  
Tesla shows the most stability in its debt-to-asset ratio, maintaining values around 31.9% to 32.5% across the three-year period. This indicates consistent financial leverage and balance sheet management.

**Microsoft**  
Microsoft's ratio peaked at 35.8% in 2022 but declined to 32.9% in 2024, indicating a trend toward lower reliance on debt and stronger equity positioning.

**Insight**  
All three companies maintain debt levels below 40% of their total assets, reflecting a conservative approach to financial leverage. Among them, Tesla demonstrates the most consistent balance sheet structure, while Microsoft shows active efforts to reduce debt dependency over time.



### Comparative Summary

| Metric             | Apple               | Tesla              | Microsoft           |
|--------------------|---------------------|---------------------|---------------------|
| Revenue Trend      | Decline → Mild Recovery | Surge → Flat     | Consistent Growth   |
| Net Income Trend   | Declining           | Spike → Collapse    | Dip → Rebound       |
| Cash Flow          | High and Stable     | Low and Flat        | Increasing           |
| Risk Profile       | Moderate            | High                | Low                 |
| Average Revenue    | 389.5 billion       | 91.9 billion        | 218.4 billion       |
